rfcs/20190729-keras-preprocessing-redesign.md

```diff
@@ -0,0 +1,351 @@
+# Keras Preprocessing API
+
+| Status        | Proposed      |
+:-------------- |:---------------------------------------------------- |
+| **Author(s)** | Francois Chollet (fchollet@google.com), Frederic Branchaud-Charron (Frederic.Branchaud-Charron@usherbrooke.ca)|
+| **Updated**   | 2019-08-21                                           |
+
+
+## Context
+
+`tf.data.Dataset` is the main API for data loading and preprocessing in TensorFLow. It has two advantages:
+
+- It supports GPU prefetching
+- It supports distribution via the Distribution Strategies API
+
+Meanwhile, `keras.preprocessing` is a major API for data loading and preprocessing in Keras. It is based
+on Numpy and Scipy, and it produces instances of the `keras.utils.Sequence` class, which are finite-length,
+resettable Python generators that yield batches of data.
+
+Some features of `keras.preprocessing` are highly useful and don't have straightforward equivalents in `tf.data`
+(in particular image data augmentation and dynamic time series iteration).
+
+Ideally, the utilities in `keras.preprocessing` should be made compatible with `tf.data`.
+This presents the opportunity to improve on the existing API. In particular we don't have good support
+for image segmentation use cases today.
+
+Some features are also being supplanted by [preprocessing layers](https://github.com/keras-team/governance/blob/master/rfcs/20190502-preprocessing-layers.md), in particular text processing. 
+As a result we may want move the current API to an API similar to Layers.
+
+
+## Goals
+
+- Unify "keras.preprocessing" and the recently-introduced [Preprocessing Layers API](https://github.com/keras-team/governance/blob/master/rfcs/20190502-preprocessing-layers.md).
+- Make all features of `keras.preprocessing` compatible with `tf.data`.
+- As a by-product, add required ops to TensorFlow (`tf.image`).
+
+
+## Proposed changes at a high-level
+
+
+- Deprecate `ImagePipelineGenerator` in favor of new `ImagePipeline` class similar to a `Sequential` model.
+- Inherits from `keras.layers.PreprocessingLayer` for all image transformations.
+- Deprecate `Tokenizer` class in favor of `TextVectorization` preprocessing layer.
+- Replace `TimeseriesGenerator` with a function-based API.
+
+
+## Detailed API changes
+
+
+### ImagePipeline
+
+#### Constructor
+
+`ImagePipeline` inherits from `keras.model.Sequential` and takes a list of layers as inputs. In the future it will inherit from `PreprocessingStage`.
+
+`ImagePipeline` is a preprocessing layer that encapsulate a series of image transformations. Since some of these transformations may be trained (featurewise normalization), it exposes the method `adapt`, like all other preprocessing layers.
```

Comment:
It's not a preproc layer, right? A preprocessing stage? Do layers implement `adapt`?

Reply:
Had the same question, i think @fchollet is referring to preprocessing stage here. 

Layers do not implement `adapt`, this was a concept introduced in the preprocessing layers design and is similar to `fit`. `adapt` is the API used to train preprocessing layers. The name `fit` was not reused since the data used for training is different between this API and fit. Also i think `adapt` was originally called `update` in the processing layers design.

Reply:
Yes, it is a preprocessing stage (so by extension it is a preprocessing layer, since preprocessing stage will subclass `PreprocessingLayer`).

I describe it as a preprocessing layer specifically because it is likely that `PreprocessingStage` will not yet exist when we ship the initial version of this API, hence `ImagePipeline` would subclass `PreprocessingLayer` in its first iteration.

The method name `adapt` was the consensus result of the preprocessing layer API design review meeting (not great, but we have to settle on something).